### PR TITLE
grpc-gateway/generator: respect full package

### PR DESCRIPTION
**Pull request by glerchundi**
Rebased on top of master, done by @notbdu.

Superseeds #417.

### CODE REVIEW

**Comment by glerchundi** [comment]
Tested on our CI and it works as expected. Take into account that `file.GoPkg.Path` **cannot be empty**.

@tmc PTAL.

**Comment by glerchundi** [comment]
do you agree @notbdu? :)

**Comment by notbdu** [comment]
Fine with me. Thanks @glerchundi.

**Comment by achew22** [comment]
Wouldn't this change the output path for every file generated, not just the ones using go_package?

**Comment by glerchundi** [comment]
@achew22 yep, it will. It's a breaking change. I can refactor to switch on a parameter or whatever and keep the old behaviour but the question is, what is desired? IMHO this is what end-user expects , me included. 

I can create another PR after this one to override `go_package` based path for one that lets user choose which output folder to use.

WDYT?

**Comment by glerchundi** [comment]
@tmc, @achew22 I changed the code to keep backward compatibility.

**Comment by glerchundi** [comment]
I don't think this is related to the PR itself...Flaky tests? IWOMM...

**Comment by glerchundi** [comment]
Any chance/free-slot to look at this, @achew22, @tmc? Thanks

**Comment by achew22** [comment]
@tmc what do you think about this?

From my vantage it appears to be the correct behavior but I would like a 2nd set of eyes before we merge a breaking change. Also, once we merge this let's do a release.

@glerchundi, could you add some tests to verify the behavior and prevent regression?

**Comment by achew22** [comment]
@glerchundi, I just tagged a release which means now is a GREAT time to introduce breaking changes like this. Please update the PR with some tests and let's get it merged.

**Comment by glerchundi** [comment]
Ok, i don’t have much time but give me a day and will try to do ASAP.

**Comment by glerchundi** [comment]
@achew22 PTAL

File: protoc-gen-grpc-gateway/gengateway/generator_test.go [hunk 1]
```diff
@@ -9,7 +10,16 @@ import (
 	"github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/descriptor"
 )
 
-func TestGenerateServiceWithoutBindings(t *testing.T) {
+func newExampleFileDescriptor() *descriptor.File {
+	return newExampleFileDescriptorWithGoPkg(
+		&descriptor.GoPackage{
+			Path: "example.com/path/to/example/example.pb",
```
Comment (achew22) [inline protoc-gen-grpc-gateway/gengateway/generator_test.go]: Shouldn't the file name end in .go?

Comment (glerchundi) [reply protoc-gen-grpc-gateway/gengateway/generator_test.go]: I don't think so.

I migrated an already existing example into a customisable function without doing any change to it but as you can observe in the two cases that I've implemented i used this descriptor as a 1-1 representation of the `go_package` proto option.

I tried to touch as less as possible :)

**Comment by achew22** [comment]
I have manually verified CLAs on glerchundi@ and notbdu@. Merging

**Comment by glerchundi** [comment]
👏🏻👏🏻

**Comment by c4milo** [comment]
How can I disable this behavior? it broke my existing code :/

**Comment by glerchundi** [comment]
@c4milo can you give us more details? Example proto file which was working and not now, as well as an expected behaviour explanation.

Although it should work as it was behaving before with just explicitly specifying the same `go_package` name as the one you're using in `package`.

**Comment by c4milo** [comment]
I should have clarified a bit better. I have no issues with the generated code itself. I had code relying on the Swagger gateway generated code being placed in the same location as the code generated by `protoc-gen-go`. The piece of code relying on that broke.

**Pull request merged**